numba/tests/test_withlifting.py

```diff
@@ -825,6 +825,37 @@ def global_var():
         ):
             global_var()
 
+    def test_objmode_gv_mod_attr(self):
+        @njit
+        def modattr1():
+            with objmode(val=types.intp):
+                val = 12.3
+            return val
+
+        @njit
+        def modattr2():
+            with objmode(val=numba.types.intp):
+                val = 12.3
+            return val
+
+        for fn in (modattr1, modattr2):
+            with self.subTest(fn=str(fn)):
+                ret = fn()
+                # the result is truncated because of the intp return-type
+                self.assertIsInstance(ret, int)
+                self.assertEqual(ret, 12)
+
+    def test_objmode_gv_mod_attr_error(self):
+        @njit
+        def moderror():
+            with objmode(val=types.THIS_DOES_NOT_EXIST):
+                val = 12.3
+            return val
+        with self.assertRaises(errors.CompilerError) as raises:
+            moderror()
+        self.assertIn("Getattr cannot be resolved at compile-time",
+                      str(raises.exception))
```

Comment:
Is it worth checking that the reported getattr that can't be resolved is as expected, e.g.:

```suggestion
                      str(raises.exception))
        self.assertIn("types.THIS_DOES_NOT_EXIST):\n            ^",
                      str(raises.exception))
```

Reply:
The suggested check depends on the error reporting options; e.g. i'm seeing colorama added codes and the suggested check is failing. I don't think it's worth writing all the regex just to match it correctly given that the current error message can only be coming from the code I added.

Reply:
OK, makes sense.

Reply:
I bumped into this whilst testing:
```python
from numba import njit, objmode
import numpy as np

@njit
def moderror(pred):
    if pred:
        x = 'intp'
    else:
        x = 'float32'
    with objmode(val=x):
        val = 12.3
    return val

moderror(False)
```
not necessarily to be fixed here, but https://github.com/numba/numba/blob/d89d34105cc0e1dc9402b14bbc0e4a7abf3c0a3e/numba/core/transforms.py#L382 should guard for multiply defined variables.